README
NAME
    Redis::Dump - It's a simple way to dump and backup data from
    redis-server

VERSION
    version 0.015

SYNOPSIS
        use Redis::Dump;
        use Data::Dumper;

        my $dump = Redis::Dump->new({ server => '127.0.0.6379', filter => 'foo' });

        print Dumper( \$dump->run );

DESCRIPTION
    It's a simple way to dump data from redis-server in JSON format or any
    format you want.

COMMAND LINE API
    This class uses MooseX::Getopt to provide a command line api. The
    command line options map to the class attributes.

METHODS
  new_with_options
    Provided by MooseX::Getopt. Parses attributes init args from @ARGV.

  run
    Perfomas the actual dump.

ATTRIBUTES
  server
    Host:Port of redis server, example: 127.0.0.1:6379.

  filter
    String to filter keys stored in redis server.

  type
    If you want to get just some types of keys.

    It can be: lists, sets, hashs, strings, zsets

  showtype
    If you want to show type with key name.

  hidevalues
    Hide value of keys.

DEVELOPMENT
    Redis::Dump is a open source project for everyone to participate. The
    code repository is located on github. Feel free to send a bug report, a
    pull request, or a beer.

    <http://www.github.com/maluco/Redis-Dump>

SEE ALSO
    Redis::Dump::Restore, App::Redis::Dump, App::Redis::Dump::Restore

SUPPORT
    You can find documentation for this module with the perldoc command.

    perldoc Redis::Dump

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Redis-Dump>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Redis-Dump>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Redis-Dump>

    *   Search CPAN

        <http://search.cpan.org/dist/Redis-Dump>

AUTHOR
    Thiago Rondon <thiago@nsms.com.br>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2011 by Thiago Rondon.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision History for Redis-Dump
===============================

0.015
-----

commit 264c8105deb46e3aebab01faa9341f666195cb87
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    eol.

commit 33da9bfceb0079db6ca2e25b6e794912796d4e5c
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Prepare for 0.15 release.

commit d6367b8dc302c29c617e05563ad6e179651aed41
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    redis-config is a new cmd for get/restore config from a redis server.

commit 562b5f8d9cf683cfaf8577cc454131110cd70f75
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Add initial support to dump config of redis server too.

v0.014 2011-07-14
-----------------

commit 6d1e968f73238f94d25def3eea0ebfe041d9d35b
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Make get_keys better with filter.

v0.013 2011-06-20
-----------------

commit 0e09c322533ebc30b1a935a8fb064c2f996b1fcc
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    New option --hidevalues.

v0.012 2011-06-16
-----------------

commit 3285a8aabeb8dca66a4af9b1aeb1438a7e56c32a
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Typo.

commit 957092dcf39eb39a91bd98bbe9e4b299f9629212
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    typo POD.

v0.011 2011-06-16
-----------------

commit 079940373e058c7d86fa8f1f044a23b965407879
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    POD, prepare to v0.011.

commit c8118b552a3dd2e32eaf585d285ba93dec1d9876
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Add initial redis-restore.

v0.010 2011-06-16
-----------------

commit aafb1f4bfe224f02d3b5ecc4410beeb34788d792
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Bugfix in filter option with zsets and sets.

v0.009 2011-06-16
-----------------

commit 3c7070f5a95b62f6e0d9ffbc2f25832d4ec9c2e6
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Fix abstract.

v0.008 2011-06-16
-----------------

commit 4902cf73d0b99836639b344654c81be83e647a9a
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Documentation.

commit 2a222c5524d8ec57678285c7ed1317e3c4e0e138
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    typo.

commit 95e8df7789dce76821e00495eb72a70f89e7bcf6
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Test for filter (type)

v0.007 2011-06-15
-----------------

commit 2d239d27e5341245fc821b39197fd525991fd628
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Show score in sorted sets.

v0.006 2011-06-15
-----------------

commit 77312e97521b8e17270305f8cb0b13133d7f8b08
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    show_type and type filter options.

v0.005 2011-06-15
-----------------

commit f5e656609be4fa377adf9fb0e714b7797b1315c6
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    POD.

v0.004 2011-06-15
-----------------

commit 73d030014a16d8abab35e7f4849f250d269bcf7d
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Support for sorted set.

commit 6438d2104915ce115b3e4c0623d39bdb18666977
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Support for set stored.

v0.003 2011-06-15
-----------------

commit 55f14502f99d7681e45e06af710f3a14dcb7267c
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Examples.

commit e1597ac73f99b341c435bea7ec2ca74de090759b
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Add filter option.

v0.002 2011-06-15
-----------------

commit 55fa81567cb257b9bbd03a714464957ed75593e2
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    0.002 dist.ini

commit abc3906a27624751ec9f3945e94fce3c1b06420e
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Use Redis 1.904.

commit 20a83595e67f15f871b4b4ef1fa40c80c316b920
Author: Thiago Berlitz Rondon <thiago.rondon@gmail.com>

    Skip Test::SpawnRedisServer in dist.ini.

===================================
End of changes in the last 365 days
===================================


lib/Redis/Dump.pm

package Redis::Dump;

use Moose;
with 'MooseX::Getopt';

use Redis 1.904;

# ABSTRACT: It's a simple way to dump and backup data from redis-server
our $VERSION = '0.015'; # VERSION

has _conn => (
    is       => 'ro',
    isa      => 'Redis',
    init_arg => undef,
    lazy     => 1,
    default  => sub { Redis->new( server => shift->server ) }
);

sub _get_keys {
    my $self   = shift;
    my $filter = $self->filter;
    return $self->_conn->keys("*$filter*") if $filter;
    return $self->_conn->keys("*");
}

sub _get_type_and_filter {
    my ( $self, $key ) = @_;
    return if $self->has_filter and not $key =~ $self->filter;
    my $type = $self->_conn->type($key);
    return if @{ $self->type } and not grep {/^$type/} @{ $self->type };
    return $type;
}

sub _get_value {
    my ( $self, $key, $type ) = @_;
    return $self->_conn->get($key) if $type eq 'string';
    return $self->_conn->lrange( $key, 0, -1 ) if $type eq 'list';
    return $self->_conn->smembers($key) if $type eq 'set';

    if ( $type eq 'zset' ) {
        my %hash;
        my @zsets = $self->_conn->zrange( $key, 0, -1, 'withscores' );
        for ( my $loop = 0; $loop < scalar(@zsets) / 2; $loop++ ) {
            my $value = $zsets[ $loop * 2 ];
            my $score = $zsets[ ( $loop * 2 ) + 1 ];
            $hash{$score} = $value;
        }
        return [ {%hash} ];
    }

    if ( $type eq 'hash' ) {
        my %hash;
        foreach my $item ( $self->_conn->hkeys($key) ) {
            $hash{$item} = $self->_conn->hget( $key, $item );
        }
        return {%hash};
    }
}

sub _get_values_by_keys {
    my $self = shift;
    my %keys;

    foreach my $key ( $self->_get_keys ) {
        my $type = $self->_get_type_and_filter($key) or next;
        my $show_name = $key;
        $show_name .= " ($type)" if $self->showtype;
        $keys{$show_name} = $self->hidevalues ? '' : $self->_get_value( $key, $type );
    }
    return %keys;
}



sub run {
    my $self = shift;
    return $self->_get_values_by_keys;
}


has server => (
    is            => 'ro',
    isa           => 'Str',
    default       => '127.0.0.1:6379',
    documentation => 'Host:Port of redis server (ex. 127.0.0.1:6379)'
);


has filter => (
    is            => 'ro',
    isa           => 'Str',
    default       => '',
    predicate     => 'has_filter',
    documentation => 'String to filter keys stored in redis server'
);


has type => (
    is            => 'ro',
    isa           => 'ArrayRef[Str]',
    default       => sub { [] },
    predicate     => 'has_type',
    documentation => 'Show just this type of key',
);


has showtype => (
    is            => 'ro',
    isa           => 'Bool',
    default       => 0,
    documentation => 'If you want to show type with key name.'
);


has hidevalues => (
    is            => 'ro',
    isa           => 'Bool',
    default       => 0,
    documentation => 'Hide values of keys'
);

1;



=pod

=head1 NAME

Redis::Dump - It's a simple way to dump and backup data from redis-server

=head1 VERSION

version 0.015

=head1 SYNOPSIS

    use Redis::Dump;
    use Data::Dumper;

    my $dump = Redis::Dump->new({ server => '127.0.0.6379', filter => 'foo' });

    print Dumper( \$dump->run );

=head1 DESCRIPTION

It's a simple way to dump data from redis-server in JSON format or any format
you want.

=head1 COMMAND LINE API

This class uses L<MooseX::Getopt> to provide a command line api. The command line options map to the class attributes.

=head1 METHODS

=head2 new_with_options

Provided by L<MooseX::Getopt>. Parses attributes init args from @ARGV.

=head2 run

Perfomas the actual dump.

=head1 ATTRIBUTES

=head2 server

Host:Port of redis server, example: 127.0.0.1:6379.

=head2 filter

String to filter keys stored in redis server.

=head2 type

If you want to get just some types of keys.

It can be: lists, sets, hashs, strings, zsets

=head2 showtype

If you want to show type with key name.

=head2 hidevalues

Hide value of keys.

=head1 DEVELOPMENT

Redis::Dump is a open source project for everyone to participate. The code repository
is located on github. Feel free to send a bug report, a pull request, or a
beer.

L<http://www.github.com/maluco/Redis-Dump>

=head1 SEE ALSO

L<Redis::Dump::Restore>, L<App::Redis::Dump>, L<App::Redis::Dump::Restore>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

perldoc Redis::Dump

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Redis-Dump>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Redis-Dump>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Redis-Dump>

=item * Search CPAN

L<http://search.cpan.org/dist/Redis-Dump>

=back

=head1 AUTHOR

Thiago Rondon <thiago@nsms.com.br>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Thiago Rondon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__





lib/Redis/Dump/Config.pm

package Redis::Dump::Config;

use Moose;
use MooseX::Types::Path::Class;
with 'MooseX::Getopt';

use JSON;
use Redis 1.904;

# ABSTRACT: It's a simple way to dump and backup config from redis-server
our $VERSION = '0.015'; # VERSION

has _conn => (
    is       => 'ro',
    isa      => 'Redis',
    init_arg => undef,
    lazy     => 1,
    default  => sub { Redis->new( server => shift->server ) }
);

sub _get_config {
    my $self = shift;
    my %cf;
    my $filter = $self->filter;
    my @configs = $self->_conn->config( 'get', $filter ? "*$filter*" : '*' );
    for ( my $loop = 0; $loop < scalar(@configs) / 2; $loop++ ) {
        my $name = $configs[ $loop * 2 ];
        my $value = $configs[ ( $loop * 2 ) + 1 ] || '';
        $cf{$name} = $value;
    }

    return %cf;
}

sub _restore {
    my $self = shift;
    my %keys = %{ from_json( $self->restore->slurp ) };
    foreach my $key ( keys %keys ) {
        my $value = $keys{$key} || "";
        warn "config set $key $value";
        $self->_conn->config( 'set', $key, $value );
    }
}


sub run {
    my $self = shift;

    $self->_restore if $self->has_restore;

    return $self->_get_config;
}


has server => (
    is            => 'ro',
    isa           => 'Str',
    default       => '127.0.0.1:6379',
    documentation => 'Host:Port of redis server (ex. 127.0.0.1:6379)'
);


has filter => (
    is            => 'ro',
    isa           => 'Str',
    default       => '',
    predicate     => 'has_filter',
    documentation => 'String to filter keys stored in redis server'
);


has restore => (
    is            => 'ro',
    isa           => 'Path::Class::File',
    coerce        => 1,
    predicate     => 'has_restore',
    documentation => 'Restore a config redis server to the server',
);

1;



=pod

=head1 NAME

Redis::Dump::Config - It's a simple way to dump and backup config from redis-server

=head1 VERSION

version 0.015

=head1 SYNOPSIS

    use Redis::Dump::Config;
    use Data::Dumper;

    my $dump = Redis::Dump::Config->new({ server => '127.0.0.6379', filter => 'foo' });

    print Dumper( \$dump->run );

=head1 DESCRIPTION

It's a simple way to dump config from redis-server in JSON format or any format
you want.

=head1 COMMAND LINE API

This class uses L<MooseX::Getopt> to provide a command line api. The command line options map to the class attributes.

=head1 METHODS

=head2 new_with_options

Provided by L<MooseX::Getopt>. Parses attributes init args from @ARGV.

=head2 run

Perfomas the actual dump.

=head1 ATTRIBUTES

=head2 server

Host:Port of redis server, example: 127.0.0.1:6379.

=head2 filter

String to filter keys stored in redis server.

=head2 restore

Restore a config redis server to the server.

=head1 DEVELOPMENT

Redis::Dump is a open source project for everyone to participate. The code repository
is located on github. Feel free to send a bug report, a pull request, or a
beer.

L<http://www.github.com/maluco/Redis-Dump>

=head1 SEE ALSO

L<Redis::Dump::Restore>, L<App::Redis::Dump>, L<App::Redis::Dump::Restore>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

perldoc Redis::Dump

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Redis-Dump>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Redis-Dump>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Redis-Dump>

=item * Search CPAN

L<http://search.cpan.org/dist/Redis-Dump>

=back

=head1 AUTHOR

Thiago Rondon <thiago@nsms.com.br>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Thiago Rondon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__





lib/Redis/Dump/Restore.pm

package Redis::Dump::Restore;

use Moose;
use MooseX::Types::Path::Class;
use JSON;
with 'MooseX::Getopt';

use Redis 1.904;

# ABSTRACT: It's a simple way to restore data to redis-server based on redis-dump.
our $VERSION = '0.014'; # VERSION

has _conn => (
    is       => 'ro',
    isa      => 'Redis',
    init_arg => undef,
    lazy     => 1,
    default  => sub { Redis->new( server => shift->server ) }
);

sub _set_string {
    my ( $self, $name, $value ) = @_;
    $self->_conn->set( $name => $value );
}

sub _set_array {
    my ( $self, $name, $value ) = @_;

    foreach my $item ( @{$value} ) {
        my $type = ref($item);
        $self->_conn->rpush( $name, $item ) if !$type;    # list

        if ( $type eq 'HASH' ) {
            my %zsets = %{$item};
            foreach my $range ( keys %zsets ) {
                $self->_conn->zadd( $name, $range, $zsets{$range} );
            }
        }

    }

}

sub _set_hash {
    my ( $self, $name, $value ) = @_;

    my %sets = %{$value};
    foreach my $item ( keys %sets ) {
        my $type = ref($item);
        $self->sadd( $name, $item, $sets{$item} ) if !$type;    # smembers

        if ( $type eq 'HASH' ) {
            my %hashs = %{$item};
            foreach my $key ( keys %hashs ) {
                $self->_conn->hset( $name, $key, $hashs{$key} );
            }
        }
    }
}

sub _set_values_by_keys {
    my $self = shift;

    $self->_conn->flushall if $self->flushall;

    my %keys = %{ from_json( $self->file->slurp ) };
    foreach my $key ( keys %keys ) {
        my $type = ref( $keys{$key} );
        $self->_set_string( $key, $keys{$key} ) if !$type;
        $self->_set_array( $key, $keys{$key} ) if $type eq 'ARRAY';
        $self->_set_hash( $key, $keys{$key} ) if $type eq 'HASH';
    }
    return 1;
}


sub run {
    my $self = shift;

    my $fh = $self->file->openr() or confess "Can't read file: $!";

    return $self->_set_values_by_keys;
}


has server => (
    is            => 'ro',
    isa           => 'Str',
    default       => '127.0.0.1:6379',
    documentation => 'Host:Port of redis server (ex. 127.0.0.1:6379)'
);


has flushall => (
    is            => 'ro',
    isa           => 'Bool',
    default       => 0,
    documentation => 'Remove all keys from all databases'
);


has file => (
    is            => 'ro',
    isa           => 'Path::Class::File',
    coerce        => 1,
    documentation => 'File with restore'
);

1;



=pod

=head1 NAME

Redis::Dump::Restore - It's a simple way to restore data to redis-server based on redis-dump.

=head1 VERSION

version 0.014

=head1 SYNOPSIS

    use Redis::Dump::Restore;

    my $restore = Redis::Dump::Restore->new({ server => '127.0.0.6379', file => 'foo.txt');

=head1 DESCRIPTION

It's a simple way to restore data to redis-server based on redis-dump (JSON).

=head1 COMMAND LINE API

This class uses L<MooseX::Getopt> to provide a command line api. The command line options map to the class attributes.

=head1 METHODS

=head2 new_with_options

Provided by L<MooseX::Getopt>. Parses attributes init args from @ARGV.

=head2 run

Perfomas the actual restore.

=head1 ATTRIBUTES

=head2 server

Host:Port of redis server, example: 127.0.0.1:6379.

=head2 flushall

Remove all keys from all databases

=head2 file

File with dump

=head1 DEVELOPMENT

Redis::Dump is a open source project for everyone to participate. The code repository
is located on github. Feel free to send a bug report, a pull request, or a
beer.

L<http://www.github.com/maluco/Redis-Dump>

=head1 SEE ALSO

L<Redis::Dump>, L<App::Redis::Dump>, L<App::Redis::Dump::Restore>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

perldoc Redis::Dump

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Redis-Dump>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Redis-Dump>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Redis-Dump>

=item * Search CPAN

L<http://search.cpan.org/dist/Redis-Dump>

=back

=head1 AUTHOR

Thiago Rondon <thiago@nsms.com.br>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Thiago Rondon.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__





Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "It's a simple way to dump and backup data from redis-server",
  "AUTHOR" => "Thiago Rondon <thiago\@nsms.com.br>",
  "BUILD_REQUIRES" => {
    "English" => 0,
    "Exporter" => 0,
    "File::Find" => 0,
    "File::Temp" => 0,
    "IO::String" => 0,
    "POSIX" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "base" => 0,
    "warnings" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Redis-Dump",
  "EXE_FILES" => [
    "bin/redis-config",
    "bin/redis-dump",
    "bin/redis-restore",
    "bin/redis-config",
    "bin/redis-dump",
    "bin/redis-restore"
  ],
  "LICENSE" => "perl",
  "NAME" => "Redis::Dump",
  "PREREQ_PM" => {
    "JSON" => 0,
    "Moose" => 0,
    "MooseX::Getopt" => 0,
    "MooseX::Types::Path::Class" => 0,
    "Redis" => "1.904",
    "strict" => 0
  },
  "VERSION" => "0.015",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



